File: move_explain.py
"""Explain one move using only adjacent public game snapshots."""

SUITS = {"H": "hearts", "D": "diamonds", "C": "clubs", "S": "spades"}
RANKS = {"J": "jack", "Q": "queen", "K": "king", "A": "ace"}


def explain_move(before, after, perspective=0):
    """Return the last move's public result, from one fixed player's perspective."""
    if before is None or before == after:
        return None
    actor = before["turn"]
    drawn = [max(0, after["hand_counts"][p] - before["hand_counts"][p]) for p in (0, 1)]
    names = ["You" if p == perspective else "Your opponent" for p in (0, 1)]
    possessives = ["Your" if p == perspective else "Your opponent’s" for p in (0, 1)]
    detail = []
    requested = 0
    recipient = None

    def card_name(card):
        return f"{RANKS.get(card[:-1], card[:-1])} of {SUITS[card[-1]]}"

    def draw_text(player):
        count = drawn[player]
        return f"{names[player]} drew {count} card{'s' if count != 1 else ''}."

    if after["phase"] == "response" and after["hand_counts"][actor] < before["hand_counts"][actor]:
        kind = "play"
        title = f"{names[actor]} declared {card_name(after['top'])}."
        if before["phase"] == "response":
            detail.append(f"Accepted {card_name(before['top'])} and played face down.")
        if after["chosen_suit"]:
            detail.append(f"Continue with {SUITS[after['chosen_suit']]}, if accepted.")
        if not after["hand_counts"][actor]:
            detail.append(f"{possessives[actor]} hand is empty; the result is still pending.")
        elif not detail:
            detail.append("Played face down; awaiting a response.")
    elif before["phase"] == "response" and after["phase"] in ("turn", "finished"):
        if after["top_status"] == "Revealed":
            caught = after["top"] != before["top"]
            kind = "bluff_caught" if caught else "challenge_failed"
            title = (f"{names[actor]} caught a bluff." if caught else
                     f"{possessives[actor]} challenge was wrong.")
            recipient = 1 - actor if caught else actor
            requested = before["draw_penalty"] + 2
            detail.append(draw_text(recipient))
        elif after["top_status"] == "Accepted" and drawn[actor] and before["hand_counts"][actor]:
            kind = "draw"
            recipient, requested = actor, before["draw_penalty"] or 1
            title = draw_text(actor)
            detail.append(f"Accepted {card_name(before['top'])}; the turn ended.")
        elif after["top_status"] == "Accepted":
            kind = "accept"
            title = f"{names[actor]} accepted {card_name(before['top'])}."
            if not before["hand_counts"][actor] and before["draw_penalty"]:
                recipient, requested = actor, before["draw_penalty"]
                detail.append(draw_text(actor))
            elif before["skip_pending"] and any(before["hand_counts"]):
                detail.append(f"{possessives[actor]} turn was skipped under the ace.")
        else:
            return None
    elif before["phase"] == "turn" and after["phase"] in ("turn", "finished"):
        if any(drawn):
            kind = "draw"
            recipient = next(p for p in (0, 1) if drawn[p])
            requested = before["draw_penalty"] or 1
            title = draw_text(recipient)
            detail.append(f"{possessives[actor]} turn ended.")
        elif before["skip_pending"] and not after["skip_pending"] and before["hand_counts"] == after["hand_counts"]:
            kind = "skip"
            title = f"{names[actor]} skipped the turn."
            detail.append("The ace’s skip is cleared.")
        else:
            return None
    else:
        return None

    if recipient is not None and drawn[recipient] < requested:
        detail.append("No more cards were available; the unpaid penalty was cancelled.")
    if after["winner"] is not None:
        winner = after["winner"]
        detail.append("You win." if winner == perspective else "Your opponent wins.")
    elif recipient is not None and not before["hand_counts"][recipient] and drawn[recipient]:
        detail.append(f"{names[recipient]} returned to play.")
    return {"kind": kind, "actor": actor, "drawn": drawn, "winner": after["winner"],
            "title": title, "detail": " ".join(detail)}

File: src-rust/src/move_explain.rs
//! Public explanations derived only from adjacent public snapshots.
use crate::game::{Card, Suit};
use serde::Deserialize;
use serde_json::{Value, json};

pub fn suit_name(suit: Suit) -> &'static str {
    match suit {
        Suit::H => "hearts",
        Suit::D => "diamonds",
        Suit::C => "clubs",
        Suit::S => "spades",
    }
}

pub fn card_name(card: Card) -> String {
    let rank = ["7", "8", "9", "10", "jack", "queen", "king", "ace"][card.rank() as usize];
    format!("{rank} of {}", suit_name(card.suit()))
}

#[derive(Deserialize)]
struct Snapshot {
    phase: String,
    turn: usize,
    hand_counts: [usize; 2],
    top: Card,
    chosen_suit: Option<Suit>,
    draw_penalty: u32,
    skip_pending: bool,
    winner: Option<usize>,
    top_status: String,
}

/// The default UI perspective is player zero. Hidden cards never enter this function.
pub fn explain_move(before: Option<&Value>, after: &Value, perspective: usize) -> Option<Value> {
    let before = before?;
    if before == after {
        return None;
    }
    let before: Snapshot = serde_json::from_value(before.clone()).ok()?;
    let after: Snapshot = serde_json::from_value(after.clone()).ok()?;
    let actor = before.turn;
    if actor > 1 || perspective > 1 {
        return None;
    }
    let drawn = [
        after.hand_counts[0].saturating_sub(before.hand_counts[0]),
        after.hand_counts[1].saturating_sub(before.hand_counts[1]),
    ];
    let names = [
        if perspective == 0 {
            "You"
        } else {
            "Your opponent"
        },
        if perspective == 1 {
            "You"
        } else {
            "Your opponent"
        },
    ];
    let possessives = [
        if perspective == 0 {
            "Your"
        } else {
            "Your opponent’s"
        },
        if perspective == 1 {
            "Your"
        } else {
            "Your opponent’s"
        },
    ];
    let draw_text = |player: usize| {
        format!(
            "{} drew {} card{}.",
            names[player],
            drawn[player],
            if drawn[player] == 1 { "" } else { "s" }
        )
    };
    let mut detail = Vec::new();
    let mut requested = 0;
    let mut recipient = None;
    let kind;
    let title;
    if after.phase == "response" && after.hand_counts[actor] < before.hand_counts[actor] {
        kind = "play";
        title = format!("{} declared {}.", names[actor], card_name(after.top));
        if before.phase == "response" {
            detail.push(format!(
                "Accepted {} and played face down.",
                card_name(before.top)
            ));
        }
        if let Some(suit) = after.chosen_suit {
            detail.push(format!("Continue with {}, if accepted.", suit_name(suit)));
        }
        if after.hand_counts[actor] == 0 {
            detail.push(format!(
                "{} hand is empty; the result is still pending.",
                possessives[actor]
            ));
        } else if detail.is_empty() {
            detail.push("Played face down; awaiting a response.".into());
        }
    } else if before.phase == "response" && matches!(after.phase.as_str(), "turn" | "finished") {
        if after.top_status == "Revealed" {
            let caught = after.top != before.top;
            kind = if caught {
                "bluff_caught"
            } else {
                "challenge_failed"
            };
            title = if caught {
                format!("{} caught a bluff.", names[actor])
            } else {
                format!("{} challenge was wrong.", possessives[actor])
            };
            let player = if caught { 1 - actor } else { actor };
            recipient = Some(player);
            requested = before.draw_penalty as usize + 2;
            detail.push(draw_text(player));
        } else if after.top_status == "Accepted"
            && drawn[actor] > 0
            && before.hand_counts[actor] > 0
        {
            kind = "draw";
            recipient = Some(actor);
            requested = before.draw_penalty.max(1) as usize;
            title = draw_text(actor);
            detail.push(format!(
                "Accepted {}; the turn ended.",
                card_name(before.top)
            ));
        } else if after.top_status == "Accepted" {
            kind = "accept";
            title = format!("{} accepted {}.", names[actor], card_name(before.top));
            if before.hand_counts[actor] == 0 && before.draw_penalty > 0 {
                recipient = Some(actor);
                requested = before.draw_penalty as usize;
                detail.push(draw_text(actor));
            } else if before.skip_pending && before.hand_counts.iter().any(|&count| count > 0) {
                detail.push(format!(
                    "{} turn was skipped under the ace.",
                    possessives[actor]
                ));
            }
        } else {
            return None;
        }
    } else if before.phase == "turn" && matches!(after.phase.as_str(), "turn" | "finished") {
        if let Some(player) = drawn.iter().position(|&count| count > 0) {
            kind = "draw";
            recipient = Some(player);
            requested = before.draw_penalty.max(1) as usize;
            title = draw_text(player);
            detail.push(format!("{} turn ended.", possessives[actor]));
        } else if before.skip_pending
            && !after.skip_pending
            && before.hand_counts == after.hand_counts
        {
            kind = "skip";
            title = format!("{} skipped the turn.", names[actor]);
            detail.push("The ace’s skip is cleared.".into());
        } else {
            return None;
        }
    } else {
        return None;
    }
    if recipient.is_some_and(|player| drawn[player] < requested) {
        detail.push("No more cards were available; the unpaid penalty was cancelled.".into());
    }
    if let Some(winner) = after.winner {
        detail.push(
            if winner == perspective {
                "You win."
            } else {
                "Your opponent wins."
            }
            .into(),
        );
    } else if let Some(player) =
        recipient.filter(|&player| before.hand_counts[player] == 0 && drawn[player] > 0)
    {
        detail.push(format!("{} returned to play.", names[player]));
    }
    Some(
        json!({"kind": kind, "actor": actor, "drawn": drawn, "winner": after.winner, "title": title, "detail": detail.join(" ")}),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    fn snapshot() -> Value {
        json!({"phase":"response","turn":1,"hand_counts":[1,2],"top":"7H","chosen_suit":null,"draw_penalty":6,"skip_pending":false,"winner":null,"top_status":"Awaiting response"})
    }
    #[test]
    fn challenge_shortage_and_winner_text_match_public_contract() {
        let before = snapshot();
        let mut after = before.clone();
        after["phase"] = json!("finished");
        after["top"] = json!("JC");
        after["top_status"] = json!("Revealed");
        after["hand_counts"] = json!([2, 2]);
        after["winner"] = json!(1);
        let result = explain_move(Some(&before), &after, 0).unwrap();
        assert_eq!(
            result,
            json!({"kind":"bluff_caught","actor":1,"drawn":[1,0],"winner":1,"title":"Your opponent caught a bluff.","detail":"You drew 1 card. No more cards were available; the unpaid penalty was cancelled. Your opponent wins."})
        );
        assert!(explain_move(None, &after, 0).is_none());
        assert!(explain_move(Some(&before), &before, 0).is_none());
    }
    #[test]
    fn queen_direct_response_uses_public_declaration_and_fixed_perspective() {
        let before = snapshot();
        let mut after = before.clone();
        after["hand_counts"] = json!([1, 1]);
        after["top"] = json!("QS");
        after["chosen_suit"] = json!("C");
        let result = explain_move(Some(&before), &after, 1).unwrap();
        assert_eq!(result["title"], "You declared queen of spades.");
        assert_eq!(
            result["detail"],
            "Accepted 7 of hearts and played face down. Continue with clubs, if accepted."
        );
    }
}
